File: apina-java/src/test/java/fi/evident/apina/java/reader/JavaTypeMatchers.java
package fi.evident.apina.java.reader;

import fi.evident.apina.java.model.type.*;
import org.hamcrest.BaseMatcher;
import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

import java.util.List;

import static java.util.Arrays.asList;

/**
 * Hamcrest matches for {@link JavaType}s.
 */
final class JavaTypeMatchers {

    public static Matcher<JavaType> basicType(Class<?> cl) {
        return basicType(new JavaBasicType(cl));
    }

    public static Matcher<JavaType> basicType(JavaBasicType type) {
        return new JavaTypeMatcher() {

            @Override
            protected boolean matchBasicType(JavaBasicType item) {
                return item.equals(type);
            }

            @Override
            public void describeTo(Description description) {
                description.appendValue(type);
            }
        };
    }

    public static Matcher<JavaType> arrayType(Matcher<JavaType> elementTypeMatcher) {
        return new JavaTypeMatcher() {

            @Override
            protected boolean matchArrayType(JavaArrayType item) {
                return elementTypeMatcher.matches(item.getElementType());
            }

            @Override
            public void describeTo(Description description) {
                description.appendDescriptionOf(elementTypeMatcher).appendText("[]");
            }
        };
    }

    @SafeVarargs
    public static Matcher<JavaType> typeVariable(String name, Matcher<JavaType>... boundMatchers) {
        return typeVariable(name, asList(boundMatchers));
    }

    private static Matcher<JavaType> typeVariable(String name, List<Matcher<JavaType>> boundMatchers) {
        return new JavaTypeMatcher() {

            @Override
            protected boolean matchTypeVariable(JavaTypeVariable item) {
                return name.equals(item.getName())
                        && matchList(item.getBounds(), boundMatchers);
            }

            @Override
            public void describeTo(Description description) {
                description.appendText("type variable ").appendValue(name).appendText(" with " + boundMatchers.size() + " bounds");
            }
        };
    }

    @SafeVarargs
    public static Matcher<JavaType> genericType(Class<?> base, Matcher<JavaType>... args) {
        return genericType(basicType(base), asList(args));
    }

    public static Matcher<JavaType> genericType(Matcher<JavaType> baseMatcher, List<Matcher<JavaType>> argMatchers) {
        return new JavaTypeMatcher() {

            @Override
            protected boolean matchParameterizedType(JavaParameterizedType item) {
                return baseMatcher.matches(item.getBaseType())
                        && matchList(item.getArguments(), argMatchers);
            }

            @Override
            public void describeTo(Description description) {
                description.appendDescriptionOf(baseMatcher).appendList("<", ",", ">", argMatchers);
            }
        };
    }

    public static Matcher<? super JavaType> typeWithRepresentation(String typeRepresentation) {
        return new TypeSafeMatcher<JavaType>() {
            @Override
            protected boolean matchesSafely(JavaType item) {
                return typeRepresentation.equals(item.toString());
            }

            @Override
            public void describeTo(Description description) {
                description.appendText("type represented by ").appendValue(typeRepresentation);
            }
        };
    }

    private static boolean matchList(List<JavaType> items, List<Matcher<JavaType>> matchers) {
        if (matchers.size() != items.size())
            return false;

        for (int i = 0, size = items.size(); i < size; i++)
            if (!matchers.get(i).matches(items.get(i)))
                return false;

        return true;
    }

    private static abstract class JavaTypeMatcher extends BaseMatcher<JavaType> {

        @Override
        public boolean matches(Object item) {
            if (item instanceof JavaArrayType)
                return matchArrayType((JavaArrayType) item);
            else if (item instanceof JavaBasicType)
                return matchBasicType((JavaBasicType) item);
            else if (item instanceof JavaTypeVariable)
                return matchTypeVariable((JavaTypeVariable) item);
            else if (item instanceof JavaParameterizedType)
                return matchParameterizedType((JavaParameterizedType) item);
            else
                return false;
        }

        protected boolean matchBasicType(JavaBasicType item) {
            return false;
        }

        protected boolean matchTypeVariable(JavaTypeVariable item) {
            return false;
        }

        protected boolean matchArrayType(JavaArrayType item) {
            return false;
        }

        protected boolean matchParameterizedType(JavaParameterizedType item) {
            return false;
        }
    }
}

File: apina-core/src/test/kotlin/fi/evident/apina/java/reader/JavaTypeMatchers.kt
package fi.evident.apina.java.reader

import fi.evident.apina.java.model.type.JavaType
import fi.evident.apina.java.model.type.TypeSchema
import org.hamcrest.BaseMatcher
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.TypeSafeMatcher
import java.util.Arrays.asList
import java.util.Collections.singletonMap

/**
 * Hamcrest matches for [JavaType]s.
 */
internal object JavaTypeMatchers {

    fun basicType(cl: Class<*>): Matcher<JavaType> = basicType(JavaType.Basic(cl))

    fun basicType(type: JavaType.Basic): Matcher<JavaType> = object : JavaTypeMatcher() {

        override fun matchBasicType(item: JavaType.Basic) = item == type

        override fun describeTo(description: Description) {
            description.appendValue(type)
        }
    }

    fun arrayType(elementTypeMatcher: Matcher<JavaType>): Matcher<JavaType> = object : JavaTypeMatcher() {

        override fun matchArrayType(item: JavaType.Array) = elementTypeMatcher.matches(item.elementType)

        override fun describeTo(description: Description) {
            description.appendDescriptionOf(elementTypeMatcher).appendText("[]")
        }
    }

    fun typeVariable(name: String): Matcher<JavaType> = object : JavaTypeMatcher() {

        override fun matchTypeVariable(item: JavaType.Variable) = name == item.name

        override fun describeTo(description: Description) {
            description.appendText("type variable ").appendValue(name)
        }
    }

    fun genericType(base: Class<*>, vararg args: Matcher<JavaType>): Matcher<JavaType> =
        genericType(basicType(base), args.asList())

    fun genericType(baseMatcher: Matcher<JavaType>, argMatchers: List<Matcher<JavaType>>): Matcher<JavaType> = object : JavaTypeMatcher() {

        override fun matchParameterizedType(item: JavaType.Parameterized) =
            baseMatcher.matches(item.baseType) && matchList(item.arguments, argMatchers)

        override fun describeTo(description: Description) {
            description.appendDescriptionOf(baseMatcher).appendList("<", ",", ">", argMatchers)
        }
    }

    fun typeWithRepresentation(typeRepresentation: String): Matcher<in JavaType> = object : TypeSafeMatcher<JavaType>() {
        override fun matchesSafely(item: JavaType) = typeRepresentation == item.toString()

        override fun describeTo(description: Description) {
            description.appendText("type represented by ").appendValue(typeRepresentation)
        }
    }

    fun singletonSchema(variable: String, vararg types: Matcher<JavaType>): Matcher<TypeSchema> =
            schema(singletonMap(JavaType.Variable(variable), types.asList()))

    private fun schema(map: Map<JavaType.Variable, List<Matcher<JavaType>>>): Matcher<TypeSchema> = object : TypeSafeMatcher<TypeSchema>() {
        override fun matchesSafely(item: TypeSchema): Boolean {
            val variables = item.variables

            if (variables.size != map.size || !map.keys.containsAll(variables))
                return false

            for ((key, value) in map)
                if (!matchList(item.getTypeBounds(key), value))
                    return false

            return true
        }

        override fun describeTo(description: Description) {
            description.appendValue(map)
        }
    }

    private fun matchList(items: List<JavaType>, matchers: List<Matcher<JavaType>>) =
            items.size == matchers.size && items.indices.all { i -> matchers[i].matches(items[i]) }

    private abstract class JavaTypeMatcher : BaseMatcher<JavaType>() {

        override fun matches(item: Any) = when (item) {
            is JavaType.Array -> matchArrayType(item)
            is JavaType.Basic -> matchBasicType(item)
            is JavaType.Variable -> matchTypeVariable(item)
            is JavaType.Parameterized -> matchParameterizedType(item)
            else -> false
        }

        protected open fun matchBasicType(item: JavaType.Basic): Boolean = false
        protected open fun matchTypeVariable(item: JavaType.Variable): Boolean = false
        protected open fun matchArrayType(item: JavaType.Array): Boolean = false
        protected open fun matchParameterizedType(item: JavaType.Parameterized): Boolean = false
    }
}
